consensus/src/pipeline/pipeline_builder.rs

```diff
@@ -377,11 +380,15 @@ impl PipelineBuilder {
 
     /// Precondition: Block is inserted into block tree (all ancestors are available)
     /// What it does: Wait for all data becomes available and verify transaction signatures
-    async fn prepare(preparer: Arc<BlockPreparer>, block: Arc<Block>) -> TaskResult<PrepareResult> {
+    async fn prepare(
+        preparer: Arc<BlockPreparer>,
+        block: Arc<Block>,
+        qc: Option<Arc<QuorumCert>>,
+    ) -> TaskResult<PrepareResult> {
         let _tracker = Tracker::new("prepare", &block);
         // the loop can only be abort by the caller
         let input_txns = loop {
-            match preparer.prepare_block(&block).await {
+            match preparer.prepare_block(&block, qc.clone()).await {
```

Comment:
this probably does't work, since the future is built when we receive the block, we need to pass the Arc of Mutex here and check if qc is set on every loop run